### PR TITLE
Implement SuperfluousCurlyBraceFinder for rule S6396

### PR DESCRIPTION
**Pull request by stanislav-jordanov-sonarsource**
Related to SONARPHP-1272

### CODE REVIEW

File: regex-parsing/src/main/java/org/sonarsource/analyzer/commons/regex/finders/SuperfluousCurlyBraceFinder.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    Quantifier quantifier = tree.getQuantifier();
+    int min = quantifier.getMinimumRepetitions();
+    Integer max = quantifier.getMaximumRepetitions();
+
+    if (max != null && max == min) { // is it (xyz){N,N}
+      if (min == 1) { // is it (xyz){1}
+        super.visitRepetition(tree);
+        regexElementIssueReporter.report(quantifier, "Remove this unnecessary quantifier.", null, Collections.emptyList());
+      } else if (min == 0) {  // is it (xyz){0}
+        regexElementIssueReporter.report(tree, "Remove this unnecessarily quantified expression.", null, Collections.emptyList());
```
Comment (nils-werner-sonarsource) [inline regex-parsing/src/main/java/org/sonarsource/analyzer/commons/regex/finders/SuperfluousCurlyBraceFinder.java:51]: Please us static Strings instead inlined messages.

File: regex-parsing/src/main/java/org/sonarsource/analyzer/commons/regex/finders/SuperfluousCurlyBraceFinder.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  }
+
+  @Override
+  public void visitRepetition(RepetitionTree tree) {
+
+    Quantifier quantifier = tree.getQuantifier();
+    int min = quantifier.getMinimumRepetitions();
+    Integer max = quantifier.getMaximumRepetitions();
+
+    if (max != null && max == min) { // is it (xyz){N,N}
```
Comment (nils-werner-sonarsource) [inline regex-parsing/src/main/java/org/sonarsource/analyzer/commons/regex/finders/SuperfluousCurlyBraceFinder.java:46]: This comment does not match `{0}` as example

File: regex-parsing/src/test/resources/finders/SuperfluousCurlyBraceFinder.yml [hunk 3]
```diff
@@ -0,0 +1,14 @@
+- '(abc){1}' # Noncompliant {{Remove this unnecessary quantifier.}}
+- '(abc){1,1}' # Noncompliant {{Remove this unnecessary quantifier.}}
```
Comment (nils-werner-sonarsource) [inline regex-parsing/src/test/resources/finders/SuperfluousCurlyBraceFinder.yml:2]: What about  `(abc){4,4}` I think this rule should also include these cases. WDYT?

**Review by nils-werner-sonarsource — CHANGES_REQUESTED** [review]
Please fix code smells, and see comments. All in all LGTM.

File: regex-parsing/src/test/resources/finders/SuperfluousCurlyBraceFinder.yml (hunk 3 above)
Comment (stanislav-jordanov-sonarsource) [reply regex-parsing/src/test/resources/finders/SuperfluousCurlyBraceFinder.yml:2]: Discussed on slack. We agreed that this looks like a legit rule, but not like one named "SuperfluousCurlyBraceFinder" ...

**Review by nils-werner-sonarsource — APPROVED** [review]
LGTM!

**Pull request merged**